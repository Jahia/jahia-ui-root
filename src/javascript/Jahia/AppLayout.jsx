import React, {useState} from 'react';
import {registry} from '@jahia/registry';
import {Route, Switch} from 'react-router';
import classnames from 'classnames';
import css from './AppLayout.scss';
import {Typography} from '@jahia/moonstone'

const AppLayout = () => {
    const routes = registry.find({type: 'route', target: 'jahia'});
    const links = routes.map(r => r.defaultPath);

    const [open, setOpen] = useState(false);

    return (
        <div className={css.AppLayout}>
            <div className={classnames(css.AppLayout_nav, {[css.open]: open, [css.close]: !open})}>


            </div>
            <div className={css.AppLayout_main}>
                <Typography variant="page">Page</Typography>

                <Switch>
                    {routes.map(r =>
                        <Route key={r.key} path={r.path} render={r.render}/>
                    )}
                </Switch>
            </div>
        </div>
    );
};

export default AppLayout;
