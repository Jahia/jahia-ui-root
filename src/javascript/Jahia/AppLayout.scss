$open-width: 250px;
$close-width: 50px;

.AppLayout {
    display: flex;
    flex-direction: row;
    height: 100vh;

    &_nav {
        overflow: hidden;

        &.open {
            flex: 0 0 $open-width;
        }

        &.close {
            flex: 0 0 $close-width;
        }
    }

    &_main {
        flex: 1 1 0;
        display: flex
    }
}
