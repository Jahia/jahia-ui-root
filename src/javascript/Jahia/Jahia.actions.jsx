function actions(registry) {
}

export default actions;
