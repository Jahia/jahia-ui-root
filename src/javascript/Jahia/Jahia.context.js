import {createContext} from 'react';

const JahiaContext = createContext({});

export default JahiaContext;
