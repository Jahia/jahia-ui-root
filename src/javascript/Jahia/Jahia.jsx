import React from 'react';
import PropTypes from 'prop-types';
import {Route, Switch} from 'react-router';
import {GlobalStyle, LayoutApp, PrimaryNav, Star} from '@jahia/moonstone';
import JahiaLogo from '../JahiaLogo';
import {useNodeInfo} from '@jahia/data-helper';
import {shallowEqual, useSelector} from 'react-redux';
import {LoginCheck} from './LoginCheck';
import {Error404, ErrorBoundary, LoaderSuspense, SiteSwitching} from '../shared';

const Jahia = ({routes, topNavGroups, bottomNavGroups}) => {
    const current = useSelector(state => ({language: state.language, site: state.site}), shallowEqual);
    const requiredPermission = ['jcr:read_default'];
    const requiredPaths = [];
    routes.filter(route => route.requiredPermission !== undefined).forEach(route => {
        if (!requiredPermission.includes(route.requiredPermission)) {
            requiredPermission.push(route.requiredPermission);
            if (route.requiredPermissionPath === undefined) {
                requiredPaths.push(`/sites/${current.site}`);
            }
        }

        if (route.requiredPermissionPath !== undefined && !requiredPaths.includes(route.requiredPermissionPath)) {
            requiredPaths.push(route.requiredPermissionPath);
        }
    });
    if (requiredPaths.length === 0) {
        requiredPaths.push(`/sites/${current.site}`);
    }

    const {loading, nodes, error} = useNodeInfo({
        paths: requiredPaths, language: current.language
    }, {getPermissions: requiredPermission}, {fetchPolicy: 'no-cache', nextFetchPolicy: 'no-cache'});

    if (error) {
        console.error('Jahia - An error occur while getting permissions', error);
        return null;
    }

    if (loading || !nodes) {
        // Wait for the query to be done.
        return null;
    }

    let primaryNav;
    if (window.contextJsParameters.config.environment === '') {
        primaryNav = (
            <PrimaryNav
                headerLogo={<JahiaLogo/>}
                top={topNavGroups}
                bottom={bottomNavGroups}
            />
        );
    } else {
        primaryNav = (
            <PrimaryNav
                headerLogo={<JahiaLogo/>}
                headerCaption={`${window.contextJsParameters.config.environment}`}
                modeIcon={<Star/>}
                top={topNavGroups}
                bottom={bottomNavGroups}
            />
        );
    }

    const filteredRoutes = routes.filter(route => {
        if (!route.requiredPermission) {
            return true;
        }

        const permissionNode = nodes.find(node => {
            if (route.requiredPermissionPath !== undefined) {
                return node.path === route.requiredPermissionPath;
            }

            return node.path === `/sites/${current.site}`;
        });

        return permissionNode && permissionNode[route.requiredPermission];
    });

    if (!nodes.every(node => {
        return requiredPaths.find(path => path === node.path) !== undefined;
    })) {
        console.error('Jahia - An error occur while getting permissions, GraphQL nodes do not match required paths', nodes, requiredPaths);
        return (
            <>
                <GlobalStyle/>
                <LoginCheck/>
                <LayoutApp
                    navigation={primaryNav}
                    content={
                        <LoaderSuspense>
                            <Switch>
                                <Route path="*"
                                       component={SiteSwitching}/>
                            </Switch>
                        </LoaderSuspense>
                    }
                />
            </>
        );
    }

    return (
        <>
            <GlobalStyle/>
            <LoginCheck/>
            <LayoutApp
                navigation={primaryNav}
                content={
                    <LoaderSuspense>
                        <Switch>
                            {filteredRoutes.map(r => (
                                <Route key={r.key}
                                       path={r.path}
                                       render={p => (<ErrorBoundary>{r.render(p)}</ErrorBoundary>)}
                            />
))}
                            <Route path="*" component={Error404}/>
                        </Switch>
                    </LoaderSuspense>
}
            />
        </>
    );
};

Jahia.propTypes = {
    routes: PropTypes.array, topNavGroups: PropTypes.array, bottomNavGroups: PropTypes.array
};

export default Jahia;
