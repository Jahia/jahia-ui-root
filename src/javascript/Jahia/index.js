import JahiaContainer from './Jahia.container';

export default JahiaContainer;
