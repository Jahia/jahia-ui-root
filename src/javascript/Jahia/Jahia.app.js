// Register redux
import React from 'react';
import {ReduxProvider} from './ReduxProvider';
import {batchDispatchMiddleware} from 'redux-batched-actions';
import thunk from 'redux-thunk';
import {ConnectedRouter, connectRouter, routerMiddleware} from 'connected-react-router';
import JahiaContext from './Jahia.context';
import Jahia from './Jahia';
import PrimaryNavGroup from './PrimaryNavGroup';
import {createBrowserHistory} from 'history';

export const jahiaApps = (registry, jahiaCtx) => {
    registry.add('app', 'redux', {
        targets: ['root:1'],
        render: next => (<ReduxProvider jahiaCtx={jahiaCtx}>{next}</ReduxProvider>)
    });
    registry.add('redux-middleware', 'batch', {middleware: batchDispatchMiddleware});
    registry.add('redux-middleware', 'thunk', {middleware: thunk});

    // Connected router
    const history = createBrowserHistory({basename: jahiaCtx.contextPath + jahiaCtx.urlbase});
    registry.add('app', 'router', {
        targets: ['root:2'],
        render: next => <ConnectedRouter history={history}>{next}</ConnectedRouter>
    });
    registry.add('redux-reducer', 'router', {targets: ['root'], reducer: connectRouter(history)});
    registry.add('redux-middleware', 'router', {middleware: routerMiddleware(history)});

    // Jahia Context
    registry.add('app', 'jahiacontext', {
        targets: ['root:3'],
        render: next => <JahiaContext.Provider value={jahiaCtx}>{next}</JahiaContext.Provider>
    });

    // UI-root
    registry.add('app', 'jahia', {
        targets: ['root:99'], render: next => (
            <>
                <Jahia routes={registry.find({type: 'route', target: 'nav-root-top'})}
                       topNavGroups={[
                           <PrimaryNavGroup key="workflowNavGroup"
                                            navItems={registry.find({
                                                type: 'workflowGroup',
                                                target: 'nav-root-workflow'
                                            })}/>,
                           <PrimaryNavGroup key="topNavGroup"
                                            navItems={registry.find({
                                                type: 'topNavGroup',
                                                target: 'nav-root-top'
                                            })}/>
                       ]}
                       bottomNavGroups={[
                           <PrimaryNavGroup key="bottomDocumentationGroup"
                                            isDisplayedWhenCollapsed={false}
                                            navItems={registry.find({
                                                type: 'bottomDocumentationGroup',
                                                target: 'nav-root-bottom'
                                            })}/>,
                           <PrimaryNavGroup key="bottomProfileGroup"
                                            navItems={registry.find({
                                                type: 'bottomProfileGroup',
                                                target: 'nav-root-bottom'
                                            })}/>,
                           <PrimaryNavGroup key="bottomAdminGroup"
                                            navItems={registry.find({
                                                type: 'bottomAdminGroup',
                                                target: 'nav-root-bottom'
                                            })}/>
                       ]}
                />
                {next}
            </>
        )
    });
};
