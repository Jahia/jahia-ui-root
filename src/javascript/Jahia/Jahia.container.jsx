import PropTypes from 'prop-types';
import {registry} from '@jahia/ui-extender';
import './Jahia.routes';
import './Workflow/Workflow.routes';
import './PageComposer/PageComposer.routes';
import {jahiaApps} from './Jahia.app';
import {jahiaRedux} from './Jahia.redux';

const JahiaContainer = ({jahiaCtx}) => {
    jahiaApps(registry, jahiaCtx);
    jahiaRedux(registry, jahiaCtx);

    // App shell
    const apps = registry.find({type: 'app', target: 'root'}).map(m => m.render);
    const render = apps.reduceRight((prevFn, nextFn) => (...args) => nextFn(prevFn(...args)), value => value);

    return render();
};

JahiaContainer.propTypes = {
    jahiaCtx: PropTypes.object
};

export default JahiaContainer;
