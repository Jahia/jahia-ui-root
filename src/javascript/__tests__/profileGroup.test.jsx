import React from 'react';
import renderer from 'react-test-renderer';
import {ProfileGroup} from '../Jahia/Jahia.routes';

describe('profile group', () => {

    test('profiles', () => {
        window.contextJsParameters = { user: {fullname: "mock_fullname", email: "mock_email"}};
        const component = renderer.create(
            <ProfileGroup />,
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
});
