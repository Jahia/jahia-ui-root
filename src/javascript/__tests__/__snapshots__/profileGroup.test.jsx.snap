// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profile group profiles 1`] = `
<li
  className="_38IL5xx6MzHEv52EctnPV_"
>
  <div
    className="_3Wbxt6soOkNE807wmyu_iI"
  >
    <div
      className="_1Ai3EiPwY-pZNUwN6mrrmu"
    >
      <svg
        className=" moonstone-icon moonstone-icon_big"
        fill="none"
        size="big"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      className="_3ejht2paDjYw7X613AENsu"
    >
      <span
        className="_2UHrI1QWqxLhuY_tZNdPZo _3Ywjur7x0TjWis06g69HVo _1Bzmq62UqnegYx7zkkgnjt _3McV6_vYIoTTSdX0ayC7p5"
      >
        jahia-ui-root:primaryNavigation.profile.label
      </span>
      <div
        className="_2UHrI1QWqxLhuY_tZNdPZo _2opCufopnd5EvablhQZyqn _1Bzmq62UqnegYx7zkkgnjt _3-oKLpu-6QNsN6e8au5YaP _3McV6_vYIoTTSdX0ayC7p5"
      >
        mock_fullname (mock_email)
      </div>
    </div>
  </div>
  <div
    className="ZgcqhWSs--AqdmVMuLrBi"
  >
    <button
      className="_3_Ozh1xIkutWoTIuxPhs4N _14jOdJL0WEJyoXAifg4_hD _3aK2Qwc8q_XoRQQWKTdY_c _3WWZMmQF_b6Rp1IO-yNNHv _1LQ77fynycbnHQ463dWcSm _2w6j1jrniCARwwHovk9Ycr"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <svg
        className=" moonstone-icon moonstone-icon_default"
        fill="none"
        size="default"
        viewBox="0 0 24 24"
      >
        <path
          d="M13 3h-2v10h2V3zm4.83 2.17l-1.42 1.42A6.92 6.92 0 0119 12c0 3.87-3.13 7-7 7A6.995 6.995 0 017.58 6.58L6.17 5.17A8.932 8.932 0 003 12a9 9 0 0018 0c0-2.74-1.23-5.18-3.17-6.83z"
          fill="currentColor"
        />
      </svg>
      <span
        className="_2UHrI1QWqxLhuY_tZNdPZo _3KgbY3KNkL4q4ieIxAbetO _3McV6_vYIoTTSdX0ayC7p5"
      >
        jahia-ui-root:primaryNavigation.profile.signOut
      </span>
    </button>
  </div>
</li>
`;
